storage/pin/pin.go

```diff
@@ -0,0 +1,494 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package pin
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"sync"
+
+	"github.com/ethersphere/swarm/api"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+)
+
+const (
+	Version        = "1.0"
+	WorkerChanSize = 8             // Max no of goroutines when walking the file tree
+	HeaderName     = "x-swarm-pin" // Presence of this in header indicates pinning required
+)
+
+var (
+	errInvalidChunkData = errors.New("invalid chunk data")
+)
+
+// FileInfo is the struct that stores the information about pinned files
+// A map of this is stored in the state DB
+type FileInfo struct {
+	isRaw      bool
+	fileSize   uint64
+	pinCounter uint64
+}
+
+// MarshalBinary encodes the FileInfo object in to a binary form for storage
+func (f *FileInfo) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 17)
+	if f.isRaw {
+		data[0] = 1
+	} else {
+		data[0] = 0
+	}
+	binary.BigEndian.PutUint64(data[1:], f.fileSize)
+	binary.BigEndian.PutUint64(data[9:], f.pinCounter)
+	return data, nil
+}
+
+// UnmarshalBinary decodes the binary form from the state store to the FileInfo object
+func (f *FileInfo) UnmarshalBinary(data []byte) error {
```

Comment:
It would be nice to add a check if data is at least 17 bytes long and return an error instead allowing to panic.

Reply:
ok. Added.